[IMP] website: allow the user to load more themes

loco-odoo opened the pull request:
[IMP] website: allow the user to load more themes

This commit adds a 'View more Themes' button in the "Theme" selection
page of the website configurator. If clicked, more themes are loaded and
the user can choose from 27 of them.

This commit also solves a small visual problem and keeps the size of the
odoo logo when more themes are loaded.

Steps to reproduce:
- Create a new website and go on the theme selection screen.
- Click on the "View more themes" button.

-> The odoo logo becomes very small.

IAP PR: https://github.com/odoo/iap-apps/pull/889
task-4184418


--------------------------------------------------------------------------------------------------------------------------------------------------------

[IMP] website: adjust configurator preview scroll animation

Prior to this commit, the preview scrolling animation used fixed values.
But since the theme redesign, image previews now have different heights.
Depending on this height, the scrolling animation may display an empty
space or not go all the way to the end of the screenshot.
This commit adjusts the scrolling animation to match the height of
selected theme previews.

task-4177975

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/180346.png)](https://mergebot.odoo.com/odoo/odoo/pull/180346)

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -442,47 +443,101 @@ export class FeaturesSelectionScreen extends ApplyConfiguratorScreen {
     }
 }
 
+export class ExtraThemesSelection extends ApplyConfiguratorScreen {
+    static template = "website.Configurator.ExtraThemesSelection";
+    static props = {
+        chooseTheme: Function,
+        blockUiDuringImageLoading: Function,
+    };
+    setup() {
+        super.setup();
+        this.state = useStore();
+        this.extraThemeSVGPreviews = [];
+        for (let i = 0; i < this.state.extraThemes.length; i++) {
+            this.extraThemeSVGPreviews.push(useRef(`ExtraThemePreview${i}`));
+        }
+
+        onMounted(() => {
+            this.props.blockUiDuringImageLoading(this.state.extraThemes, this.extraThemeSVGPreviews);
+        });
+    }
+}
```
As discussed: `useEffect` can probably avoid this extra component class.

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -442,47 +443,101 @@ export class FeaturesSelectionScreen extends ApplyConfiguratorScreen {
+
+    async getMoreThemes() {
+        const removeLoadingEffect = addButtonLoadingEffect(this.extraThemesButtonRef.el);
+        const themes = await this.orm.call("website",
+            "configurator_recommended_themes",
```
To share somehow, at minimum in an utility function in this file.

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -442,47 +443,101 @@ export class FeaturesSelectionScreen extends ApplyConfiguratorScreen {
+        const extraThemes = themes.filter((extraTheme) => !mainThemeNames.includes(extraTheme.name));
+        this.state.updateExtraThemes(extraThemes, this.nbrExtraThemes);
+        removeLoadingEffect();
+        if (!this.state.extraThemes.length) {
+            // Hide the extra themes button as no extra themes have been found
+            this.extraThemesButtonRef.el.classList.add("d-none");
+        }
```
That info should be in the state, so that this button is not even rendered in the case no extra theme was found

loco-odoo replied:
Adapted but I am not sure if the way I did is the best: Indeed, I modified `env.store` to add the extra `showExtraThemesButton` element to the state. I did it this way as `env.store` is a proxy (and so, I can not do something like `Object.assign({}, env.store, {'showExtraThemesButton':true})`). The problem is that I modify the store so components that use the store see `showExtraThemesButton` that is only relevant for the `ThemeSelectionScreen` component.

qsm-odoo reviewed (changes requested):
First batch of review :)

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.xml:
```diff
@@ -193,7 +193,7 @@
 <t t-name="website.Configurator.ThemeSelectionScreen">
     <div class="o_configurator_screen h-100 d-flex flex-column o_theme_selection_screen">
-        <div class="o_configurator_odoo_logo mt-4 ms-5" style="height: 31px; width: 99px;"/>
+        <div class="o_configurator_odoo_logo flex-shrink-0 mt-4 ms-5" style="height: 31px; width: 99px;"/>
```
You can squash this @loco-odoo

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -480,15 +480,18 @@ def configurator_init(self):
         return r
 
     @api.model
-    def configurator_recommended_themes(self, industry_id, palette):
```
Unless I am misunderstanding something:
```
Steps to reproduce:
- Create a new website and go on the theme selection screen.
- Click on the "View more themes" button.

-> The odoo logo becomes very small.
```
... there is no "view more themes" button before this very commit, so this is not a fix... just part of the new implementation. Just remove the steps to reproduce.

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -82,6 +84,33 @@ export const PALETTE_NAMES = [
 export const CUSTOM_BG_COLOR_ATTRS = ["menu", "footer"];
 
+const NBR_MAIN_THEMES = 3;
+const NBR_EXTRA_THEMES = 24;
```
Should this be hardcoded? Should not this be "number of available themes"?

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -82,6 +84,33 @@ export const PALETTE_NAMES = [
+ * theme names and their related text svgs (as result of a Promise). The length
+ * of the list is at most 'resultNbrMax'.
+ */
+async function getRecommendedThemes(orm, industryId, palette, resultNbrMax=NBR_MAIN_THEMES) {
```
Like this, I agree that this function has low value (why not call the ORM directly repeating the python function). To have value, pass the whole state instead of industryId and palette. Then you can read the state and then this function has value: it reads the correct entry from the state and call the python function accordingly.

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.js:
```diff
@@ -447,42 +469,82 @@ export class ThemeSelectionScreen extends ApplyConfiguratorScreen {
+    /**
+     * Transforms text svgs into svg elements and adds a loading effect that
+     * blocks the UI during the loading of the images inside those svg elements.
+     * @param {Array<Object>} themes - The text svgs.
```
Blank line before `@param`

qsm-odoo commented on addons/website/static/src/client_actions/configurator/configurator.xml:
```diff
@@ -231,6 +231,29 @@
                             </t>
                         </div>
                     </div>
+                    <t t-if="state.showExtraThemesButton and !state.extraThemes.length and state.themes.length == nbrMainThemes">
```
I am not sure to understand the logic here: why is not `!state.extraThemes.length` enough? Do we actually need `showExtraThemesButton`? And why the `state.themes.length == nbrMainThemes` ?

loco-odoo replied:
- The `showExtraThemesButton` is for the edge case where there are only 3 themes to show. In this case, if the user clicks on the button, no extra themes are shown (`state.extraThemes.length ==0`) so I thought it would be better to hide the button in this case -> The user tries to load more themes but there are no available, no need to retry.
- The `state.themes.length == nbrMainThemes` is for a case where there are less than 3 available themes. In this case, it though it would be better to not even show the button as we already know that clicking on it would have no effect :+1:

qsm-odoo replied:
Create a property in your component which returns this condition and where you can add some documentation about this :+1:

qsm-odoo reviewed (changes requested):
@loco-odoo Priority on this :pray:

loco-odoo commented:
@qsm-odoo I adapted according to our discussion :+1: Note that in the end, I found a better condition to display the 'view more themes' button; In the end, I show it if there is enough main themes and if the button was not clicked already :+1: 
Note that there is still [this](https://github.com/odoo/odoo/pull/180346#discussion_r1764688321) (not sure if you saw it) (it is now `extraThemesLoaded` instead of `showExtraThemesButton`) but the problem is the same.

qsm-odoo commented:
> Note that there is still https://github.com/odoo/odoo/pull/180346#discussion_r1764688321

Yes, I saw, as long as it works fine by me for now :)

qsm-odoo reviewed (changes requested):
@loco-odoo 

Bug:
- Configurator -> Load more themes
- Browser: Back -> Back -> Forward -> Forward => Crash

Also, there is a strange loading effect when loading all themes:
1. It takes a long time (the button is spinning for a long time) -> is that normal? Is the RPC that slow?
2. Once the button stops spinning, the screen is loaded immediately but there is a flicker block UI which pops up for a micro seconds. Any way to improve that please?

loco-odoo commented:
@qsm-odoo 
> Bug:

Fixed :+1: I changed the code a bit; `extraThemes` is not part of the `Store` anymore as it is only something that is useful for the `ThemeSelectionScreen` component. 

> Once the button stops spinning, the screen is loaded immediately but there is a flicker block UI

I fixed it by delaying a bit the blocking of the UI :+1: 

> It takes a long time (the button is spinning for a long time) -> is that normal? Is the RPC that slow?

Indeed, it is the rpc that is quite slow. In fact, I checked a bit and it appears that it is the processing part of the rpc (`_process_svg()` of `website.configurator.feature`) that takes a lot of time; in local, for 27 images, it takes already 3 seconds. I guess that for 100 svg, it could become a bit problematic. Maybe would it be better if we load them 24 by 24 with the 'View more themes' button at the end in order to load them little by little?

qsm-odoo commented:
@loco-odoo 

> Maybe would it be better if we load them little by little ( e.g. 24 by 24) with the 'View more themes' button at the end ?

Maybe, or maybe have the route stream the svg one by one as they arrive. Or maybe 3 by 3 just by scrolling. We'll see later, that's fine for now.

> Fixed 👍

Ok I'll review, test and merge if ok :+1: 

> I checked a bit and it appears that it is the processing part of the rpc (`_process_svg()` of `website.configurator.feature`) that takes a lot of time

Maybe you can investigate that a bit more (what part of process_svg takes the most time? why? can it be improved?). But don't push here if you find a solution, that can be done after merge :+1:

qsm-odoo reviewed (approved):
@loco-odoo For the block UI issue, I finally spoke with framework-js and made a commit to improve the block ui service (first commit of the PR now). Adding a delay like you did is not what we wanted as it would leave the UI unblocked for 700ms while displaying the SVG with unloaded images (basically making the call to block ui totally useless as the point was that users were clicking too fast on theme choices without even looking at them (image not even loaded yet).

Otherwise, I retested everything, seems good :+1: 

@robodoo rebase-ff r+

> For the colors, I think we should probably stick to this code as we have to find case insensitive substring (e.g: `#3AADAA`, `#3aadaa`, etc...) in the svg and replace them by the wanted colors

We probably can enforce to use only uppercased color hex codes. Those SVG are made by us and they can only use 5 colors anyway. Would be good to enforce them uppercased either way.

> In this case, the search is not case insensitive but the gain in execution time is huge for 27 images: +/- 1.483 second currently -> +/- 0.011 seconds with the diff.

Strange, but nice if it is that easy to fix :clap: Could you make a PR about it?

robodoo commented:
Merge method set to rebase and fast-forward.

loco-odoo commented:
> Could you make a PR about it?

@qsm-odoo, sure, [here it is](https://github.com/odoo/odoo/pull/181201) :+1:
